File: .verify/device_profile.py
#!/usr/bin/env python3
"""
Reimplementa DeviceProfile e verifica o perfil de estabilidade contra um
catalogo de aparelhos reais.

O risco especifico deste recurso: ele grava configuracoes automaticamente. Um
erro aqui nao trava o app -- ele degrada silenciosamente a experiencia de quem
nunca vai reclamar, ou pior, sobrescreve o que o usuario ajustou a mao. As
verificacoes abaixo focam nesses dois pontos.
"""
import math
import sys

FAILURES = []


def check(cond, msg):
    if not cond:
        FAILURES.append(msg)


LOW, MEDIUM, HIGH = "LOW", "MEDIUM", "HIGH"


def detect_tier(ram, cores, is32, gles, sdk):
    """Espelha DeviceProfile.detectTier()."""
    if ram < 4096 or cores <= 4 or is32 or gles < 3:
        return LOW
    if ram >= 8192 and cores >= 8 and sdk >= 30:
        return HIGH
    return MEDIUM


def resolution_scale(min_side, tier, increment=5):
    """Espelha DeviceProfile.recommendedResolutionScale()."""
    target = {LOW: 540, MEDIUM: 720, HIGH: 1080}[tier]
    if min_side <= target:
        return 100
    scale = round(100 * target / min_side)
    scale = int(math.ceil(scale / increment) * increment)
    floor = 35 if tier == LOW else 50
    return max(floor, min(scale, 100))


def recommended_settings(tier):
    """Espelha o conjunto de chaves gravadas por applyRecommendedDefaults()."""
    high = tier == HIGH
    return {
        "alternate_surface":    tier != LOW,
        "sustainedPerformance": high,
        "force_vsync":          high,
        "bigCoreAffinity":      tier != HIGH,
        "checkLibraries":       tier != LOW,
        "allocationAutomatic":  True,
    }


# (nome, RAM, cores, 32bits, GLES, SDK, lado menor da tela, tier esperado)
DEVICES = [
    ("Poco X6 Pro",                 11500, 8, False, 3, 34, 1220, HIGH),
    ("Xiaomi 2311DRK48G",           11323, 8, False, 3, 36, 1080, HIGH),
    ("Galaxy S23",                   7800, 8, False, 3, 34, 1080, MEDIUM),
    ("Pixel 8",                     11800, 9, False, 3, 34, 1080, HIGH),
    ("Moto G32",                     3800, 8, False, 3, 31,  720, LOW),
    ("Redmi 9A",                     1900, 4, False, 3, 29,  720, LOW),
    ("Galaxy J5 2016",               1500, 4, True,  2, 23,  720, LOW),
    ("tablet GLES2",                 2000, 4, False, 2, 22,  800, LOW),
    ("Galaxy A54",                   7600, 8, False, 3, 33, 1080, MEDIUM),
    ("aparelho 8GB Android 9",       8192, 8, False, 3, 28, 1080, MEDIUM),
    ("aparelho 16GB gamer",         15500, 8, False, 3, 34, 1440, HIGH),
]

print(f"{'Aparelho':24s}{'Tier':>8}{'Escala':>9}   surface  sustained  vsync  bigCore")
print("-" * 78)
for name, ram, cores, is32, gles, sdk, side, expected in DEVICES:
    tier = detect_tier(ram, cores, is32, gles, sdk)
    check(tier == expected, f"{name}: classificado {tier}, esperado {expected}")

    scale = resolution_scale(side, tier)
    floor = 35 if tier == LOW else 50
    check(floor <= scale <= 100, f"{name}: escala {scale} fora de {floor}-100%")
    check(scale % 5 == 0, f"{name}: escala {scale} nao e multiplo do incremento")

    s = recommended_settings(tier)
    # Invariante central: aparelho de entrada nunca recebe ajuste caro
    if tier == LOW:
        check(not s["sustainedPerformance"],
              f"{name}: desempenho sustentado num aparelho de entrada corta o clock de pico")
        check(not s["alternate_surface"],
              f"{name}: surface alternativa em aparelho de entrada (driver antigo)")
        check(not s["checkLibraries"],
              f"{name}: verificacao SHA-1 atrasa o lancamento em aparelho fraco")
        check(scale <= 100, f"{name}: escala deveria reduzir")
    # RAM automatica e sempre ligada: e a protecao contra o Android matar o jogo
    check(s["allocationAutomatic"], f"{name}: RAM automatica deveria estar ligada")

    print(f"{name:24s}{tier:>8}{scale:>8}%   "
          f"{str(s['alternate_surface']):<9}{str(s['sustainedPerformance']):<11}"
          f"{str(s['force_vsync']):<7}{s['bigCoreAffinity']}")

# --- Monotonicidade: hardware melhor nunca deve receber perfil pior ----------
order = {LOW: 0, MEDIUM: 1, HIGH: 2}
prev = -1
for ram in range(2048, 16385, 512):
    tier = detect_tier(ram, 8, False, 3, 34)
    check(order[tier] >= prev,
          f"{ram} MB recebeu tier menor que um aparelho com menos RAM")
    prev = order[tier]

# --- Escala de resolucao: tela maior nunca deve renderizar mais pixels -------
for tier in (LOW, MEDIUM, HIGH):
    prev_pixels = None
    for side in (540, 720, 900, 1080, 1440, 2160):
        scale_pct = resolution_scale(side, tier)
        pixels = side * scale_pct / 100
        target = {LOW: 540, MEDIUM: 720, HIGH: 1080}[tier]
        floor = 35 if tier == LOW else 50
        # Quando o piso de legibilidade e atingido, nao da para chegar ao alvo:
        # a checagem entao e so que o piso foi de fato respeitado.
        at_floor = scale_pct == floor
        check(at_floor or pixels <= target * 1.12,
              f"tier {tier}, tela {side}: renderiza {pixels:.0f}px, alvo {target}px")
        prev_pixels = pixels

# --- Qualquer aparelho GLES 2 ou 32 bits e sempre LOW ------------------------
for ram in (2048, 8192, 16384):
    for cores in (4, 8):
        check(detect_tier(ram, cores, True, 3, 34) == LOW,
              f"32 bits com {ram} MB/{cores} cores nao foi classificado LOW")
        check(detect_tier(ram, cores, False, 2, 34) == LOW,
              f"GLES 2 com {ram} MB/{cores} cores nao foi classificado LOW")

# --- Preservacao das escolhas do usuario -------------------------------------
def apply(prefs, force, recommended):
    """Espelha setIfAbsent(): so grava o que o usuario nunca tocou."""
    result = dict(prefs)
    for key, value in recommended.items():
        if force or key not in prefs:
            result[key] = value
    return result

user_prefs = {"alternate_surface": False, "force_vsync": True}
after = apply(user_prefs, False, recommended_settings(HIGH))
check(after["alternate_surface"] is False,
      "aplicacao automatica sobrescreveu uma escolha explicita do usuario")
check(after["force_vsync"] is True,
      "aplicacao automatica sobrescreveu force_vsync do usuario")
check("bigCoreAffinity" in after,
      "chaves nao tocadas pelo usuario deveriam receber o padrao recomendado")

forced = apply(user_prefs, True, recommended_settings(HIGH))
check(forced["alternate_surface"] is True,
      "o botao manual deveria sobrescrever, ja que o usuario pediu")

if FAILURES:
    print("\nFALHAS:")
    for f in FAILURES:
        print("  -", f)
    sys.exit(1)
print("\nPerfil de aparelho: todas as verificacoes passaram")

File: app_pojavlauncher/src/main/java/net/kdt/pojavlaunch/utils/DeviceProfile.java
package net.kdt.pojavlaunch.utils;

import android.app.ActivityManager;
import android.content.Context;
import android.content.SharedPreferences;
import android.os.Build;
import android.util.DisplayMetrics;
import android.util.Log;

import net.kdt.pojavlaunch.Architecture;
import net.kdt.pojavlaunch.Tools;
import net.kdt.pojavlaunch.prefs.LauncherPreferences;

/**
 * Classifica o aparelho e aplica uma configuracao conhecidamente estavel para
 * a classe dele.
 * <p>
 * <b>Por que isto existe.</b> Os padroes do launcher foram calibrados em
 * aparelhos topo de linha. Num celular de entrada os mesmos padroes produzem
 * travamentos, encerramento pelo Android por pressao de memoria ou tela preta
 * -- e o usuario nao tem como saber que mexer em "surface alternativa" ou na
 * escala de resolucao resolveria. Aqui o launcher decide por ele.
 * <p>
 * <b>Principio de projeto:</b> este perfil so escolhe <i>padroes</i>. Tudo o
 * que o usuario ajustou manualmente continua valendo -- ver
 * {@link #applyRecommendedDefaults(Context, boolean)}.
 */
public final class DeviceProfile {
    private static final String TAG = "DeviceProfile";

    /** Marca que os padroes ja foram aplicados uma vez, para nao sobrescrever o usuario. */
    public static final String PREF_KEY_PROFILE_APPLIED = "deviceProfileApplied";
    /** Versao do perfil: incrementar reaplica os padroes apos uma atualizacao do app. */
    public static final String PREF_KEY_PROFILE_VERSION = "deviceProfileVersion";
    private static final int CURRENT_PROFILE_VERSION = 1;

    /** Classe de desempenho do aparelho. */
    public enum Tier {
        /**
         * Pouca RAM, CPU fraca, muitas vezes 32 bits ou GLES 2.
         * Prioriza "abrir e continuar aberto" acima de qualidade visual.
         */
        LOW,
        /** Intermediario: equilibra estabilidade e qualidade. */
        MEDIUM,
        /** Topo de linha: pode usar os recursos mais pesados. */
        HIGH
    }

    private DeviceProfile() {}

    /**
     * Classifica o aparelho por RAM, nucleos, arquitetura e GLES.
     * <p>
     * Usa varios sinais em vez de so a RAM porque eles discordam com frequencia:
     * ha celulares de entrada com 8 GB e CPU fraca, e aparelhos antigos de 4 GB
     * com CPU boa. O criterio e conservador -- na duvida, classifica para baixo,
     * porque um aparelho potente com configuracao conservadora perde um pouco de
     * qualidade, enquanto um aparelho fraco com configuracao agressiva nao roda.
     */
    public static Tier detectTier(Context context) {
        int ram = Tools.getTotalDeviceMemory(context);
        int cores = Runtime.getRuntime().availableProcessors();
        boolean is32Bit = Architecture.is32BitsDevice();

        // A versao do GLES so entra na conta se ja tiver sido consultada: criar um
        // contexto EGL custa dezenas de milissegundos e este metodo pode rodar na
        // thread principal durante a inicializacao. Quando ainda nao ha
        // informacao, os outros sinais decidem sozinhos.
        int gles = GLInfoUtils.hasInfo() ? GLInfoUtils.getGlInfo().glesMajorVersion : 3;

        // Qualquer um destes sozinho ja define aparelho de entrada.
        if (ram < 4096 || cores <= 4 || is32Bit || gles < 3) return Tier.LOW;

        // Topo de linha exige folga em todos os eixos.
        if (ram >= 8192 && cores >= 8 && Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            return Tier.HIGH;
        }
        return Tier.MEDIUM;
    }

    /**
     * Se o Android considera o aparelho com pouca memoria. Nesses aparelhos o
     * sistema encerra processos em segundo plano de forma muito mais agressiva.
     */
    public static boolean isLowRamDevice(Context context) {
        ActivityManager am = (ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);
        if (am == null) return false;
        return am.isLowRamDevice();
    }

    /**
     * Escala de resolucao recomendada, em porcentagem.
     * <p>
     * Renderizar em resolucao menor e, de longe, o ajuste com maior efeito sobre
     * fluidez e temperatura: o custo por quadro cai com o quadrado da escala.
     * Telas de celular sao pequenas o bastante para que a perda seja discreta.
     */
    public static int recommendedResolutionScale(Context context, Tier tier) {
        DisplayMetrics metrics = context.getResources().getDisplayMetrics();
        int minSide = Math.min(metrics.widthPixels, metrics.heightPixels);

        // Lado menor que a renderizacao deve mirar, por classe de aparelho.
        int targetSide;
        switch (tier) {
            case LOW:    targetSide = 540;  break;
            case MEDIUM: targetSide = 720;  break;
            default:     targetSide = 1080; break;
        }
        if (minSide <= targetSide) return 100;

        int scale = Math.round(100f * targetSide / minSide);
        // O seekbar trabalha em incrementos; arredondar evita um valor que a
        // interface nao consegue representar.
        int increment = context.getResources()
                .getInteger(net.kdt.pojavlaunch.R.integer.resolution_seekbar_increment);
        scale = (int) (Math.ceil((double) scale / increment) * increment);

        // Piso por classe de aparelho. Num celular de entrada com tela 1440p o
        // alvo de 540p exigiria 37%, e sem um piso o texto do jogo fica
        // ilegivel; mas usar 50% para todos faria justamente os aparelhos mais
        // fracos com tela densa continuarem renderizando pixels demais. O piso
        // menor so vale para a classe de entrada, onde rodar importa mais que
        // nitidez.
        int floor = tier == Tier.LOW ? 35 : 50;
        return Math.max(floor, Math.min(scale, 100));
    }

    /**
     * Aplica os padroes recomendados para este aparelho.
     * <p>
     * <b>Nao sobrescreve escolhas do usuario.</b> Cada ajuste so e gravado se a
     * chave ainda nao existir em SharedPreferences, ou seja, se o usuario nunca
     * tocou naquela opcao. Quem ja configurou o launcher do jeito que gosta nao
     * tem nada alterado.
     *
     * @param force ignora a marca de "ja aplicado" e regrava tudo, inclusive por
     *              cima do que o usuario mudou. Usado apenas pelo botao
     *              "Otimizar para este aparelho" nas configuracoes.
     * @return a classe detectada, para exibicao
     */
    public static Tier applyRecommendedDefaults(Context context, boolean force) {
        SharedPreferences prefs = LauncherPreferences.DEFAULT_PREF;
        Tier tier = detectTier(context);

        boolean alreadyApplied = prefs.getBoolean(PREF_KEY_PROFILE_APPLIED, false);
        int appliedVersion = prefs.getInt(PREF_KEY_PROFILE_VERSION, 0);
        if (alreadyApplied && appliedVersion >= CURRENT_PROFILE_VERSION && !force) {
            return tier;
        }

        SharedPreferences.Editor editor = prefs.edit();
        boolean lowRam = isLowRamDevice(context);
        boolean highTier = tier == Tier.HIGH;

        // Resolucao: o ajuste de maior impacto em fluidez e temperatura.
        setIfAbsent(prefs, editor, force, "resolutionRatio",
                recommendedResolutionScale(context, tier));

        // Surface alternativa (SurfaceView): mais estavel e economica, mas alguns
        // drivers antigos a implementam mal. So ligamos onde ha confianca.
        setIfAbsent(prefs, editor, force, "alternate_surface", tier != Tier.LOW);

        // Modo de desempenho sustentado: o Android reduz o pico de clock para
        // evitar throttling termico. Em aparelho fraco o pico e justamente o que
        // segura os quadros, entao so faz sentido no topo de linha.
        setIfAbsent(prefs, editor, force, "sustainedPerformance", highTier);

        // VSync evita tearing e economiza bateria, mas limita os quadros. Em
        // aparelho fraco o problema e nao alcancar a taxa da tela, nao passar dela.
        setIfAbsent(prefs, editor, force, "force_vsync", highTier);

        // Afinidade com os nucleos rapidos ajuda quando ha nucleos lentos que
        // segurariam a thread principal do jogo.
        setIfAbsent(prefs, editor, force, "bigCoreAffinity", tier != Tier.HIGH);

        // Verificacao de SHA-1 das bibliotecas: custa CPU e I/O a cada
        // lancamento. Em aparelho de entrada isso e um atraso perceptivel, e a
        // protecao que oferece (deteccao de download corrompido) e secundaria.
        setIfAbsent(prefs, editor, force, "checkLibraries", tier != Tier.LOW);

        // RAM automatica: essencial em aparelho com pouca memoria, onde um heap
        // fixo grande demais faz o Android encerrar o jogo.
        setIfAbsent(prefs, editor, force, "allocationAutomatic", true);

        // Em aparelhos que o proprio Android marca como low-RAM, desligar a
        // animacao de fundo do launcher reduz a pressao de memoria.
        if (lowRam) {
            setIfAbsent(prefs, editor, force, "disableGestures", false);
        }

        editor.putBoolean(PREF_KEY_PROFILE_APPLIED, true);
        editor.putInt(PREF_KEY_PROFILE_VERSION, CURRENT_PROFILE_VERSION);
        editor.apply();

        Log.i(TAG, "Perfil aplicado: " + tier + " (lowRam=" + lowRam + ")");
        return tier;
    }

    /** Grava so se a chave ainda nao existir, preservando o que o usuario ajustou. */
    private static void setIfAbsent(SharedPreferences prefs, SharedPreferences.Editor editor,
                                    boolean force, String key, boolean value) {
        if (!force && prefs.contains(key)) return;
        editor.putBoolean(key, value);
    }

    private static void setIfAbsent(SharedPreferences prefs, SharedPreferences.Editor editor,
                                    boolean force, String key, int value) {
        if (!force && prefs.contains(key)) return;
        editor.putInt(key, value);
    }

    /** Resumo legivel do aparelho, para a tela de configuracoes e para o log. */
    public static String describe(Context context, Tier tier) {
        StringBuilder builder = new StringBuilder(tier.name())
                .append(" \u2022 ").append(Tools.getTotalDeviceMemory(context)).append(" MB")
                .append(" \u2022 ").append(Runtime.getRuntime().availableProcessors())
                .append(" cores");
        // So relata a GPU se ela ja tiver sido consultada, para nao criar um
        // contexto EGL apenas por causa de um texto informativo.
        if (GLInfoUtils.hasInfo()) {
            GLInfoUtils.GLInfo info = GLInfoUtils.getGlInfo();
            builder.append(" \u2022 ").append(info.getVendorFamily())
                    .append(" GLES ").append(info.glesMajorVersion);
        }
        return builder.toString();
    }
}
